# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/picture.rs
# Commit: f3b490c076d6
# Full Hash: f3b490c076d623d6ef3bcfd297a0522d9727477c
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2020-01-23 21:42:24
# Regressor Bug: 1602643
# File Overlap Count: 1
# Description:
#   Bug 1602643 - Disable WebRender compositor dinamically for async screenshot r=gw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D59658
# ==============================================================================

diff -r 0c5fb3a8a94d -r f3b490c076d6 gfx/wr/webrender/src/picture.rs
--- a/gfx/wr/webrender/src/picture.rs	Thu Jan 23 15:35:45 2020 +0200
+++ b/gfx/wr/webrender/src/picture.rs	Thu Jan 23 04:37:28 2020 +0000
@@ -608,6 +608,8 @@
         /// What changed in the primitive that was different
         prim_compare_result: PrimitiveCompareResult,
     },
+    // The compositor type changed
+    CompositorKindChanged,
 }
 
 /// A minimal subset of Tile for debug capturing
@@ -2100,6 +2102,36 @@
             frame_state.resource_cache,
         );
 
+        // If compositor mode is changed, need to drop all incompatible tiles.
+        match (frame_context.config.compositor_kind, self.native_surface_id) {
+            (CompositorKind::Draw { .. }, Some(_)) => {
+                frame_state.composite_state.destroy_native_tiles(
+                    self.tiles.values_mut(),
+                    frame_state.resource_cache,
+                );
+                for tile in self.tiles.values_mut() {
+                    tile.surface = None;
+                    // Invalidate the entire tile to force a redraw.
+                    tile.invalidate(None, InvalidationReason::CompositorKindChanged);
+                }
+                if let Some(native_surface_id) = self.native_surface_id.take() {
+                    frame_state.resource_cache.destroy_compositor_surface(native_surface_id);
+                }
+            }
+            (CompositorKind::Native { .. }, None) => {
+                // This could hit even when compositor mode is not changed,
+                // then we need to check if there are incompatible tiles.
+                for tile in self.tiles.values_mut() {
+                    if let Some(TileSurface::Texture { descriptor: SurfaceTextureDescriptor::TextureCache { .. }, .. }) = tile.surface {
+                        tile.surface = None;
+                        // Invalidate the entire tile to force a redraw.
+                        tile.invalidate(None, InvalidationReason::CompositorKindChanged);
+                    }
+                }
+            }
+            (_, _) => {}
+        }
+
         world_culling_rect
     }
 